In [ ]:
#r "nuget: FsODE, 0.0.1"
#r "nuget: FSharp.Stats, 0.4.7"
#r "nuget: Plotly.NET.Interactive, 3.0.2"

open FsODE
open Plotly.NET
open Plotly.NET.LayoutObjects
open Plotly.NET.StyleParam
open FSharp.Stats

Installed Packages FSharp.Stats, 0.4.7 FsODE, 0.0.1 Plotly.NET.Interactive, 3.0.2

Loading extensions from `Plotly.NET.Interactive.dll`

# Tag 1 – Einführung in die Wachstumsmodellierung und Grundlagen einfacher Differentialgleichungen

In biologischen Systemen gibt es verschiedene Aspekte und Gesetzmäßigkeiten, die man modellieren kann. Ein besonders häufiger und beliebter Aspekt in der Biologie ist das Wachstum. So kann man das Wachstum von Zellen, von Geweben, von Organen und Organismen, von Populationen und von Stoffmengen untersuchen. Dazu gehört auch das negative Wachstum, auch Zerfall oder Abfall genannt. Beispiele hierfür sind der Verbrauch von Substrat durch Lebewesen in einer Umgebung oder das Eingehen (Absterben) von Pflanzenindividuen auf einer Wiese infolge von anhaltender Dürre und Trockenheit.

Am ersten Tag unseres Grundpraktikums behandeln wir grundlegende Eigenschaften von Wachstum und deren Modellierung.

## Was ist Wachstumsmodellierung?

Wissenschaftliche Modelle sind Abstraktionen der Wirklichkeit. Das Ziel ist, mit einem Modell einen (meist komplexen) Sachverhalt zu beschreiben.  
Um Wachstum zu beschreiben, verwenden wir gedankliche Modelle (_mental models_), die biologische Gesetzmäßigkeiten möglichst präzise abbilden sollen. Modellierung an sich ist in der Biologie sehr wichtig, um Veränderungen und Wechselwirkungen erklären und mehr oder weniger präzise (je nach Güte des Modells) vorhersagen zu können.

Zu den ältesten und _vermeintlich_ einfachsten Experimenten gehört die Analyse von Wachstum. Dieses kann je nach untersuchtem System durch verschiedene Parameter beschrieben werden. Steht der Organismus im Fokus, so sind Gewicht, Länge oder Größe von Interesse. Bei Populationen geht es meist um eine Zell- oder Individuenzahl pro definiertem Volumen oder Fläche (Populationsdichte).
Im Labor können verschiedene Wachstumsbedingungen künstlich erzeugt werden und damit der Einfluss jedes einzelnen Faktors – getrennt oder kombiniert – untersucht werden. Zum Beispiel können Bakterien statt auf Standardmedium mit Glucose auf Medium mit Galactose angezogen werden. Eine eventuell nötige Umwandlung von Galactose in Glucose, oder aufwendigere metabolische Reaktionen führen hier zu langsamerem Wachstum. Dies lässt sich in der Regel durch einfache Absorptionsmessungen im Photometer beweisen. Ein vermindertes Wachstum geht hier mit einer langsameren Trübung (niedrigere optische Dichte) des Mediums einher, was auf einem Graphen leicht visualisiert werden kann.

![](https://raw.githubusercontent.com/CSBiology/BIO-MBP-06-U-2/main/images/day1/01_Growth_01_small.png)  
***Abbildung 1: Beispiel für einen Wachstumsverlauf.*** _Aufzucht von Bakterien in einem Glucose- bzw. Galactose-haltigen Medium. Die y-Achse gibt die Absorption im Photometer bei einer bestimmten optischen Dichte an._

Leider kann der Wachstumsunterschied mit dem oben gezeigten Datensatz nicht quantifiziert werden. Zwar kann die zeitliche Differenz zwischen zwei Punkten gleicher optischer Dichte bestimmt werden (Kreuze in Abbildung 1), jedoch ist diese Differenz sowohl stark von den gewählten Messzeitpunkten, als auch von den vorliegenden (unbekannten) Wachstumsphasen abhängig. Außerdem verläuft die Trübung nicht linear. Das bedeutet, dass sich bei einer Zellverdopplung nicht automatisch auch die Absorption verdoppelt. Für qualitative Aussagen ist diese Methode ausreichend, für quantitative Analysen aber ungeeignet.

Wesentlich aussagekräftiger ist eine Bestimmung der absoluten Zellkonzentration eines bekannten Wachstumsmodells. Die so ermittelten Parameter können für quantitative und vergleichende Analysen von z. B. Glucose- und Galactosemedium verwendet werden. Das einfachste Wachstumsmodell wollen wir uns im Folgenden anschauen.  

____

## Exkurs: Änderungsgleichungen

Biologische Prozesse können meist mit Änderungsgleichungen (= Differentialgleichungen, kurz: DGLs) dargestellt werden. Hierbei handelt es sich um eine Beschreibung der Änderung in Abhängigkeit der sich ändernden Größe. Die exakte Lösung einer solchen Änderungsgleichung ergibt die Funktion, mit der sich der biologische Sachverhalt (hier: die Wachstumsfunktion) adäquat beschreiben lässt.

$\frac{df}{dx} = 4f(x)$, eine beispielhafte DGL

Falls Sie Ihr Schulwissen über Gleichungen, Funktionen und Ableitungen auffrischen möchten, lesen Sie [dieses Notebook](link einfügen).

___

## Einfaches Wachstumsmodell

Wachstum, also die Änderung der Zellzahl über die Zeit, ist proportional zur Zellzahl $N$. Der Proportionalitätsfaktor ist $r$. Je höher die Zellzahl, desto höher ist die Rate, mit der die Zellzahl wächst (Diese Begründung bitte kurz sacken lassen).

_Aufgabe 1.1:_ Wie sieht die beschriebene DGL aus?

a) $\frac{dN}{dt} = \frac{r}{N}$

b) $N(t) = rtN$

c) $N(t) = \frac{rN}{t}$

d) $\frac{dN}{dt} = rN$


_Antwort:_ 

_Aufgabe 1.2:_ Beweisen Sie, dass $N(t) = N_0 · e^{rt}$ eine Lösung für die oben definierte Differentialgleichung ist.

Hinweis: Leiten Sie die oben stehende geschlossene Form der Wachstumsgleichung nach $t$ ab und setzen Sie $N$ und die Ableitung in die Differenzialgleichung ein. Versuchen Sie diese Aufgabe innerhalb von 10 Minuten zu lösen. Fahren Sie ansonsten mit dem nächsten Abschnitt fort.

## Simulation von Änderungsgleichungen mit approximativen Verfahren

Selten ist eine Integration von Modellen möglich. Wie oben bewiesen ist das Integral tatsächlich eine Lösung der DGL. 
In der Biologie gibt es oftmals Fragestellungen oder Systeme, in denen eine Vielzahl von Parametern und Abhängigkeiten eine Rolle spielen und somit eine Integration nicht machbar ist. Glücklicherweise ist dies auch nicht notwendig. Denn das Wissen um das Verhalten der Steigung beschreibt den Funktionsverlauf meist ausreichend.  
In der Mathematik erarbeitet man die Lösung einer solchen Änderungsgleichung über verschiedene (mögliche) Lösungswege, jedoch sind für viele DGLs nur approximative (= Näherungs-)Lösungen möglich.

### Euler-Verfahren

Ein Beispiel für eine solche Methode ist das nach _Leonhard Euler_ benannte **Euler-Verfahren**. Es handelt sich dabei um einen iterativen Algorithmus, d. h., dass eine bestimmte Berechnung mehrmals wiederholt wird, bei jeder Wiederholung jedoch mit veränderten Startbedingungen.  
Wir werden anhand des Beispiels der Bakterienkultur das Euler-Verfahren Schritt für Schritt erklären:  
Für das Euler-Verfahren wird mind. 1 Startpunkt benötigt, z. B. $P(0|100)$, das wäre also die Bakteriendichte von $100$ zum Zeitpunkt $t = 0$ Minuten. Daher wissen wir, dass der Kurvenverlauf auf jeden Fall an der $t$-Stelle $0$ $y = 100$ sein muss. Das bedeutet:

$N(0) = 100$ bzw. $N_0 = 100$

Zudem setzen wir eine sog. Schrittweite (wir nennen sie $h$). Die sagt aus, wie weit wir bei jedem Iterationsschritt nach rechts gehen (also um wieviel sich der $t$-Wert erhöht). In unserem Beispiel sei das:

$h = 1$, also unser $t$ erhöht sich bei jedem Iterationsschritt um $1$.

Jetzt fügen wir unsere Werte in die folgende Formel ein:

$N_1 = N_0 + h · \frac{dN}{dt}$, das bedeutet: Zum Zeitpunkt $t = 1$ ist die Bakterienpopulation gewachsen und zwar um den Faktor der Änderung nach der Zeit mit der Schrittweite.

und da $\frac{dN}{dt} = rN$ ist folgt

$N_1 = N_0 + h · r · N$, wobei wir für $N$ jetzt $N_0$ einsetzen:

$N_1 = N_0 + h · r · N_0$

Alle Werte eingesetzt:

$N_1 = 100 + 1 · r · 100$, kurz: $N_1 = 100 + 100r$

$r$, das ist der Wachstumsfaktor unserer Bakterienkultur, kennen wir noch nicht. Für unser Beispiel nehmen wir mal an, $r = 0.3$.

$N_1 = 100 + 0.3 * 100$, daraus folgt: $N_1 = 130$

Dies war der erste Iterationsschritt. Wir haben jetzt errechnet, dass (nach diesem Verfahren) bei $t = 1$ der $y$-Wert (also die Bakteriendichte) bei $130$ liegt, also $P_1(1|130)$ ist. Jeder folgende Schritt bezieht sich immer auf das Ergebnis des Schrittes davor:

$N_2 = N_1 + h · r · N_1$

eingesetzt:

$N_2 = 130 + 1 · 0.3 · 130$

$N_2 = 169$, nach zwei Minuten ist also die Bakteriendichte bereits bei einem Wert von 169 angelangt.

Dies kann man nun so oft wiederholen wie man will. Meist wird vorher bereits die Anzahl ($n$) an Schritten vorgegeben, bspw. $n = 10$, dann würde man also 10 Iterationsschritte berechnen.  
**Wichtig:** Das Euler-Verfahren liefert einem keine fertige Funktion, sondern nur eine beliebige Reihe an Punkten, die _ungefähr_ (!) den Kurvenverlauf abbilden.

Auf dem Papier würde eine solche Rechnung lange dauern. In der Informatik nutzen wir daher Programme, mit deren Hilfe wir uns vom Computer die Lösung _errechnen lassen_.  
Im Codeblock unten finden Sie das Euler-Verfahren als Code und der vollständigen Berechnung unseres Beispiels:

In [ ]:
// wir setzen unseren Faktor r:
let r = 0.3

// wir setzen das t0 von unserem Punkt P0, der ja (0|100) war:
let t0 = 0.

// und das dazugehörige y0 (der Startwert unserer Bakteriendichte):
let y0 = 100.

// die Schrittweite war 1:
let h = 1.

// und wir wollen 10 Iterationsschritte berechnen:
let n = 10.

// Lassen Sie sich nicht vom Code verwirren. Wichtig ist hier nur, dass der Algorithmus dem Euler-Verfahren entspricht.
let berechneEuler x0 y0 f h n =
    [|x0 .. h .. x0 + (n / (1. / h))|]
    |> Array.scan (
        fun acc x ->
            acc + h * (f acc)
    ) y0

berechneEuler t0 y0 (fun t -> r * t) h n

index value 0 100 1 130 2 169 3 219.7 4 285.61 5 371.293 6 482.6809 7 627.48517 8 815.730721 9 1060.4499373 10 1378.58491849 11 1792.160394037

Hier das Ganze als Diagramm visualisiert:

In [ ]:
// eine Zwischenspeicherung unseres Ergebnisses:
let eulerYWerte = berechneEuler t0 y0 (fun t -> r * t) h n

eulerYWerte
|> Array.indexed
|> Chart.Point
|> Chart.withTitle "Euler-Verfahren, Schrittweite = 1, Anzahl Punkte = 10"
|> Chart.withXAxisStyle "t [min]"
|> Chart.withYAxisStyle "Bakteriendichte"

<!-- Plotly chart will be drawn inside this DIV -->

_Aufgabe 1.3.:_ Wie könnten Sie als Biolog:in einen solchen Startpunkt $P$ experimentell bestimmen?

_Antwort:_

Betrachten Sie folgende Abbildung aus der Vorlesung:

![](https://raw.githubusercontent.com/CSBiology/BIO-MBP-06-U-2/main/images/day1/01_euler1_small.png)  
***Abbildung 2: Visualisierung des Euler-Verfahrens anhand eines Beispiels.***

Wie Sie anhand von Abbildung 2 sehen, ist die rechnerischer Approximation mit dem Euler-Verfahren leider gar nicht so präzise. Man kann jedoch die Präzision erhöhen, indem man die Schrittweite verringert. 

_Aufgabe 1.4.:_ Wählen Sie eine Schrittweite von 0.1, erhöhen Sie proportional dazu die Anzahl an Punkten, die simuliert werden sollen, damit auf der x-Achse der ungefähr gleiche Zeitraum abgedeckt wird. Wie viele Punkte müssen nun simuliert werden?

In [ ]:
// neue Schrittweite hier eintragen
let h2 = 

// proportional die Anzahl der Punkte erhöhen, um wieder bis t ≈ 10 zu kommen
let n2 = 

// eine Zwischenspeicherung unseres neuen Ergebnisses:
let eulerYWerte2 = berechneEuler t0 y0 (fun t -> r * t) h2 n2

eulerYWerte2
|> Array.mapi (fun i y -> float i * h2, y)
|> Chart.Point
|> Chart.withTitle "Euler-Verfahren, Schrittweite = 0.1, Anzahl Punkte = (einzutragen)"
|> Chart.withXAxisStyle "t [min]"
|> Chart.withYAxisStyle "Bakteriendichte"

Error: input.fsx (5,1)-(5,4) parse warning Possible incorrect indentation: this token is offside of context started at position (2:1). Try indenting this token further or using standard formatting conventions.
input.fsx (5,1)-(5,4) parse warning Possible incorrect indentation: this token is offside of context started at position (2:1). Try indenting this token further or using standard formatting conventions.
input.fsx (8,1)-(8,4) parse warning Possible incorrect indentation: this token is offside of context started at position (5:1). Try indenting this token further or using standard formatting conventions.
input.fsx (8,1)-(8,4) parse warning Possible incorrect indentation: this token is offside of context started at position (5:1). Try indenting this token further or using standard formatting conventions.
input.fsx (10,1)-(10,13) parse warning Possible incorrect indentation: this token is offside of context started at position (5:1). Try indenting this token further or using standard formatting conventions.
input.fsx (10,1)-(10,13) parse warning Possible incorrect indentation: this token is offside of context started at position (5:1). Try indenting this token further or using standard formatting conventions.
input.fsx (5,1)-(5,4) parse error The block following this 'let' is unfinished. Every code block is an expression and must have a result. 'let' cannot be the final code element in a block. Consider giving this block an explicit result.

Vergleichen wir mal beide Kurven:

In [ ]:
[|
    eulerYWerte
    |> Array.indexed
    |> Chart.Point
    |> Chart.withTraceInfo "Schrittweite = 1"
    eulerYWerte2
    |> Array.mapi (fun i y -> float i * h2, y)
    |> Chart.Point
    |> Chart.withTraceInfo "Schrittweite = 0.1"
|]
|> Chart.combine
|> Chart.withTitle "Approximation nach Euler-Verfahren"
|> Chart.withXAxisStyle "t [min]"
|> Chart.withYAxisStyle "Bakteriendichte"

Error: input.fsx (6,5)-(6,17) typecheck error The value or constructor 'eulerYWerte2' is not defined. Maybe you want one of the following:
   eulerYWerte
input.fsx (7,41)-(7,43) typecheck error The value or constructor 'h2' is not defined.

Mit der niedrigeren Schrittweite sieht das doch schon viel besser aus.  
Leider ist das Euler-Verfahren relativ langsam, wenn man eine höhere Präzision zum Ziel hat. Im Folgenden werden wir das demonstrieren:

_Aufgabe 1.5.:_ Führen Sie beide folgenden Codeblocke 5× aus und notieren Sie sich die Dauer, die die Berechnung gebraucht hat. Berechnen Sie die durchschnittlich benötigte Zeit.

In [ ]:
berechneEuler t0 y0 (fun t -> r * t) h n

index value 0 100 1 130 2 169 3 219.7 4 285.61 5 371.293 6 482.6809 7 627.48517 8 815.730721 9 1060.4499373 10 1378.58491849 11 1792.160394037

In [ ]:
let h4 = 0.0000001

let n3 = 10000000.

berechneEuler t0 y0 (fun t -> r * t) h4 n3

index value 0 100 1 100.000003 2 100.0000060000001 3 100.00000900000028 4 100.00001200000055 5 100.00001500000091 6 100.00001800000136 7 100.00002100000191 8 100.00002400000254 9 100.00002700000326 10 100.00003000000407 11 100.00003300000498 12 100.00003600000596 13 100.00003900000705 14 100.00004200000822 15 100.00004500000948 16 100.00004800001084 17 100.00005100001228 18 100.0000540000138 19 100.00005700001543 (9999982 more)

_Antwort_:
- erster Codeblock: 
- zweiter Codeblock: 

### Runge-Kutta-Verfahren

Die Mathematiker _Carl Runge_ und _Wilhelm Kutta_ haben sich um 1900 daher Verfahren überlegt, die ähnlich funktionieren wie das Euler-Verfahren, aber präziser sind. Man nennt diese und alle auf diesem Prinzip basierenden Verfahren _Runge-Kutta-Verfahren_, das Euler-Verfahren gehört ebenfalls dazu (auch wenn es bereits lange vorher entstand).  
Ein Beispiel für ein solches Runge-Kutta-Verfahren ist das _Heun-Verfahren_, nach Karl Heun. Dieses funktioniert wie folgt:

Die ersten Schritte verlaufen wie beim Euler-Verfahren: Man überlegt sich die Schrittweite $h$ und die Schrittanzahl $n$ und berechnet $P_1$ wie bekannt und danach $P_2$.  
Jetzt nimmt man jedoch den Wert, den man für $P_2$ erhält und berechnet den Mittelwert aus diesem und aus $P_1$ und fügt diesen für $P_1$ ein (s. Abbildung 3). Idealerweise liegt die Steigung damit näher an der tatsächlichen Funktion. Mit jedem weiteren Punkt berechnet man das ganze Prozedere wie zuvor erneut.

![](https://raw.githubusercontent.com/CSBiology/BIO-MBP-06-U-2/main/images/day1/01_heun1_small.png)  
***Abbildung 3: Visualisierung des Heun-Verfahrens anhand eines Beispiels.***

Oben sprachen wir von der Verwendung von Programmbibliotheken. Im Rest des Praktikums nutzen wir eine solche Programmbibliothek, die verschiedene Runge-Kutta-Verfahren zur Näherung anbietet. Algorithmen wie die der Programmbibliothek, die mathematische Probleme numerisch lösen, nennt man auch _Solver_.

---

Wir wollen nun die Wachstumsfunktion simulieren.

Dies tun wir, indem wir dem Algorithmus folgende Anfangswerte geben:

$N(t_0) = 100$

$t_0 = 0$

In [ ]:
// der Rechenalgorithmus der Programmbibliothek, die wir benutzen:
let modelContext = 
    OdeContext(
        OdeSolverMethod.RK546M, // das ist das Runge-Kutta-Verfahren, das wir verwenden wollen
        OdeSolverOptions(
            StepSize = 0.42 // hier stellen wir Optionen des Solvers ein, z. B., dass die Schrittweite 0.42 sein soll
        )
    )

// Anfangswerte:
let t0 = 0.
let N0 = 100.0

// unsere DGL als einfaches Modell (SimpleModel):
let dN_dt : SimpleModel =
    fun N t ->
        let cellcount = N
        let cellcount' = r * N
        cellcount'


Im Modell `dN_dt` sind nun alle relevanten Elemente vorhanden und deren Beziehung formuliert, sodass sich dieses Modell im Folgenden simulieren lässt.

In [ ]:
// hier lassen wir den Solver die Simulierung machen
let sim_dy_dx = 
    modelContext.OdeInt(t0, N0, dN_dt)
    |> SolPoints.take 10    // der Solver soll uns 10 Datenpunkte geben
    |> SolPoints.memorize

sim_dy_dx

index x Y 0 0 [ 100 ] 1 0.42 [ 113.4282167625625 ] 2 0.999259549657388 [ 134.955898363052 ] 3 2.1772271605451294 [ 192.16186029123628 ] 4 3.8940775510006906 [ 321.6263247377607 ] 5 6.039496988444697 [ 612.1674466254168 ] 6 8.515955468393773 [ 1286.8070827545662 ] 7 11.251389834936289 [ 2923.4116361761935 ] 8 14.195179461744015 [ 7069.730547285138 ] 9 17.312471927868035 [ 18009.84524032065 ]

_Aufgabe 1.6:_ Visualisieren Sie das Modell anhand der oben zwischengespeicherten 10 Datenpunkte und beschriften Sie die x- und y-Achse.

In [ ]:
let shapeOfLine =
    LayoutObjects.Shape.init(ShapeType.Line,sim_dy_dx.[1].x,sim_dy_dx.[3].x,sim_dy_dx.[1].Y.[0],sim_dy_dx.[3].Y.[0],Line=Line.init(Dash=DrawingStyle.Dash))

sim_dy_dx
|> SolPoints.toPoints 1
|> Chart.Point
|> Chart.withShape(shapeOfLine)
|> Chart.withTitle("Simulation des einfachen Wachstumsmodells")
|> Chart.withXAxisStyle("") // hier x-Achsenbeschriftung einfügen
|> Chart.withYAxisStyle("") // hier y-Achsenbeschriftung einfügen

<!-- Plotly chart will be drawn inside this DIV -->

Bei der folgenden Aufgabe möchten wir genauer betrachten

_Aufgabe 1.7:_ Schätzen Sie grob die Steigung zum Zeitpunkt $t = 1$ ab. Entspricht sie der theoretischen Steigung aus der Differentialgleichung ($m = rN$)?
  - Berechnen Sie die Steigung wie im unteren Schema (Abbildung 4) gezeigt
    - Zoomen und hovern Sie über die Punkte um die Werte bei $t = 0.42$ und $t = 2.18$ angezeigt zu bekommen.
  - Berechnen Sie die Steigung anhand der Differentialgleichung
    - Bestimmen Sie $N$, also die Bakteriendichte an $t = 1$, also bei Minute 1
    - Wachstumsfaktor $r$ ist gegeben mit $r = 0.3$
    - Berechnen Sie die Steigung $m = rN$
  - Vergleichen Sie die Steigung durch Abschätzen mit der theoretischen Steigung der Differentialgleichung. Sind sie ähnlich?

![](https://raw.githubusercontent.com/CSBiology/BIO-MBP-06-U-2/main/images/day1/01_Growth_slope_small.png)  
***Abbildung 4: Schematische Darstellung der Steigungsberechnung mit Beispiel.***

_Aufgabe 1.8 (Expertenaufgabe):_ Formulieren Sie die Wachstumsfunktion aus Aufgabe 1.2 unten im Codeblock.

_Antwort:_

In [ ]:
let y_exact =
    fun t -> 
        N0 * exp(r * t) //aufgabe

[
    sim_dy_dx
    |> SolPoints.toPoints 1
    |> Chart.Spline 
    |> Chart.withTraceInfo("simulation");
    
    sim_dy_dx
    |> SolPoints.map (fun p -> (p.x, y_exact p.x))
    |> Chart.Point
    |> Chart.withTraceInfo("exact");

]
|> Chart.combine
|> Chart.withTitle("Abbildung 3 - Simple growth model comparison")
|> Chart.withXAxisStyle("Time")
|> Chart.withYAxisStyle("Population size")

<!-- Plotly chart will be drawn inside this DIV -->